eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/MultiPartHandler.scala

```diff
@@ -48,10 +48,50 @@ class MultiPartHandler(nodeParams: NodeParams, db: IncomingPaymentsDb, commandBu
    */
   def doHandle(paymentHash: ByteVector32): Boolean = true
 
+  /** Can be overridden to do custom pre-processing on successfully received payments. */
+  def preFulfill(success: MultiPartPaymentFSM.MultiPartPaymentSucceeded)(implicit log: LoggingAdapter): MultiPartPaymentFSM.MultiPartPaymentSucceeded = success
```

Comment:
What's the point of making this method return the `MultiPartPaymentFSM.MultiPartPaymentSucceeded`? Is it that we want the user to be able to alter this object and what would be an example of that?

Reply:
Yes I figured that if this method doesn't return anything, then it cannot influence the `doFulfill` in any way so it becomes completely useless.

A plugin could modify the `PaymentPart`s to enrich them with more data from an external DB (not sure yet if this is useful). I was wondering if we should take this further and provide the `preFulfill` with the ability to delay the `doFulfill` (if for example `preFulfill` asks the user for some confirmation and wants to wait for its input without blocking the actor).
Could we return a Future to make this possible? Or another synchronization object?

Reply:
I made `DoFulfill` a message instead of a function in https://github.com/ACINQ/eclair/pull/1324/commits/d18a5bb2072c3419eb0fe8ada8f513c18b073093.
This allows a plugin to override `preFulfill` to arbitrarily delay or change what needs to be fulfilled, or even fetch a preimage from an external source.
I think we can start with that, and then evolve the abstraction as we build plugins and see what's missing.